sycl/unittests/scheduler/HostTaskAndBarrier.cpp

```diff
@@ -0,0 +1,272 @@
+
+  MainLock.unlock();
+  QueueDevImpl->wait();
+}
+
+TEST_F(BarrierHandlingWithHostTask, BarrierKernelHostTask) {
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 0u);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), true);
+
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
+  EXPECT_EQ(KernelEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  MainLock.unlock();
+  QueueDevImpl->wait();
+}
+
+TEST_F(BarrierHandlingWithHostTask, KernelHostTaskBarrier) {
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
+  EXPECT_EQ(KernelEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 1u);
+  EXPECT_EQ(BarrierWaitList[0], HostTaskEventImpl);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), false);
+
+  MainLock.unlock();
+  QueueDevImpl->wait();
+}
+
+TEST_F(BarrierHandlingWithHostTask, KernelBarrierHostTask) {
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
+  EXPECT_EQ(KernelEventImpl->isEnqueued(), true);
+
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 0u);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
```

Comment:
I would expect to see a dependency on barrier event here (similar to `BarrierHandlingWithHostTask` test case).